nrf-hal-common/src/ppi.rs

```diff
@@ -0,0 +1,192 @@
+//! HAL interface for the PPI peripheral
+//!
+//! The Programmable Peripheral Interconnect interface allows for an autonomous interoperability
+//! between peripherals through their events and tasks. There are fixed PPI channels and fully
+//! configurable ones, fixed channels can only connect specific events to specific tasks. For fully
+//! configurable channels, it is possible to choose, via software, the event and the task that it
+//! will triggered by the event.
+//!
+//! On nRF52 devices, there is also a fork task endpoint, where the user can configure one more task
+//! to be triggered by the same event, even fixed PPI channels have a configurable fork task.
+
+use crate::target::PPI;
+
+mod sealed {
+    pub trait Channel {
+        const CH: usize;
+    }
+
+    pub trait NotFixed {}
+}
+use sealed::{Channel, NotFixed};
+
+/// Trait to represent a Programmable Peripheral Interconnect channel.
+pub trait Ppi {
+    /// Enables the channel.
+    fn enable(&mut self);
+
+    /// Disables the channel.
+    fn disable(&mut self);
+
+    #[cfg(not(feature = "51"))]
+    /// Sets the fork task that must be triggered when the configured event occurs. The user must
+    /// provide the address of the task.
+    fn set_fork_task_endpoint(&mut self, addr: u32);
+}
+
+/// Traits that extends the [Ppi](trait.Ppi.html) trait, marking a channel as fully configurable.
+pub trait ConfigurablePpi {
+    /// Sets the task that must be triggered when the configured event occurs. The user must provide
+    /// the address of the task.
+    fn set_task_endpoint(&mut self, addr: u32);
+
+    /// Sets the event that will trigger the chosen task(s). The user must provide the address of
+    /// the event.
+    fn set_event_endpoint(&mut self, addr: u32);
+}
+
+// All unsafe `ptr` calls only uses registers atomically, and only changes the resources owned by
+// the type (guaranteed by the abstraction)
+impl<P: Channel> Ppi for P {
+    fn enable(&mut self) {
+        let regs = unsafe { &*PPI::ptr() };
+        regs.chenset.write(|w| unsafe { w.bits(1 << P::CH) });
+    }
+
+    fn disable(&mut self) {
+        let regs = unsafe { &*PPI::ptr() };
+        regs.chenclr.write(|w| unsafe { w.bits(1 << P::CH) });
+    }
+
+    #[cfg(not(feature = "51"))]
+    fn set_fork_task_endpoint(&mut self, addr: u32) {
+        let regs = unsafe { &*PPI::ptr() };
+        regs.fork[P::CH].tep.write(|w| unsafe { w.bits(addr) });
+    }
+}
+
+// All unsafe `ptr` calls only uses registers atomically, and only changes the resources owned by
+// the type (guaranteed by the abstraction)
+impl<P: Channel + NotFixed> ConfigurablePpi for P {
+    fn set_task_endpoint(&mut self, addr: u32) {
+        let regs = unsafe { &*PPI::ptr() };
+        regs.ch[P::CH].tep.write(|w| unsafe { w.bits(addr) });
+    }
+
+    fn set_event_endpoint(&mut self, addr: u32) {
```

Comment:
The manual says that these registers "accept only addresses to registers from the Event/Task group". Does that mean that writing an incorrect address has no effect, or will that cause trouble?

Reply:
Hmm, I just tried here on hardware, I used another timer's CC0 register (TIMER1) to see if it would write something to it or hardfault, but nothing happened, no write nor fault, so I guess it just doesn't do anything ?

Reply:
I also tried with a RAM address now, same thing.

Reply:
It might be really tedious, but we could define an extension trait for all of the events and tasks, it would also let us get rid of the awkward `&p.RADIO.tasks_disable as *const _ as u32` dance.